features/glance.feature

```diff
@@ -0,0 +1,39 @@
+@glance
+Feature: Glance
+  As an administrator
+  I want to make sure that Glance component is working correctly
+
+  @create_kvm_image
+  Scenario: Creation of KVM image
+    Given KVM source image exists
+    And glance image does not exist
+    When I create new glance image based on jeos1
+    Then the image is listed as active
+    And this image has non-empty ID
+    And its ID can be used to show the image info
+    When I delete the image
```

Comment:
> Cucumber tests are usually self-driven from inside and capture everything in step definitions. Unfortunately, with our approach we will always have two contexts - `mkcloud` infrastructure setup and "dumb" `cucumber` tests where the knowledge about the current setup is mandatory for the person who runs the tests. But this does not limit us for how to build the scenarios.

Agreed.

> I would not put constraints on what is actual testing and what not. If the scenario fails while the image is being deleted, it's straight forward to see it, more effort is actually not needed compared to a separate scenario having failed, the reason will be the same.
> 
> > I disagree - the problem is that it breaks the test suite report, because if something breaks with image deletion, any scenario which has image deletion at the end will be reported as failing.  That's bad, because a) it makes the situation look worse than it actually is, and b) extra effort is required to check all failing scenarios and ascertain that they're all only due to a single issue with image deletion.
> 
> Maybe we both mean different things.

Maybe ;-)

> I don't see the action of deletion as a generic event that is equal with all scenarios you need to check for.

Agreed.  Sometimes deletion is the thing you are focussing on testing, and other times it is just part of setup and/or teardown.

> I went to [cucumber mailing list](https://groups.google.com/forum/#!forum/cukes) to get some information from those who designed this tool and found out that I was wrong with suggesting explicit image deletion at the end of a scenario. The recommended way here is to do the cleanup in the `Given` steps.

That sounds like a good approach, yes.

> It's important to add here, that we don't do any generic cleanup, we just need to get rid of a specific file if it exists.

That's not necessarily sufficient.  For example if you had a bunch of scenarios which all created new images with different names, each `Given the image "$NAME" does not exist` would be different because `$NAME` would be different from each.  So then you would also need each scenario to have another step `Given there is enough unused quota for image "$NAME"`, otherwise there would be a risk of running out of image quota due to lack of image deletion at the end of each scenario.

Yes this is a slightly artificial example, but it would become a lot more real if we were talking about guest VM instances in nova rather than images in glance.

However the good news is that this could be done very nicely via [backgrounds](https://github.com/cucumber/cucumber/wiki/Background)!

> This is not (easily) doable in a separate scenario or hook due to scenario specific variables/config.

I don't see why not - sounds easily doable to me using a hook.

> Additionally, I would like to avoid using hooks as it could end up messy
> 
> > Why?
> 
> Every scenario would expect new hook definition.

Really?  Couldn't you simply reuse the same image cleanup hook across all scenarios which create images?

    After('@creates_kvm_image') do
      # Delete the image which was created
      control_node.openstack.image.delete @image_id
    end

> As these hooks are defined in separate files in `features/support/some_file.rb` we would need to keep that always with sync with the scenarios and features

Sure, but that is already true of all steps.  That's just the way Cucumber works.

> give these hooks some names, maybe create categories etc;

I don't think hooks need names or categories.  It's enough to associate them with tags.

> The problem is with the specific config and variables for the scenarios - the hooks apply mostly for general things that are not dependent on any things defined within a scenario.

I don't see the problem - doesn't the use of @image_id in the above hook take care of that?

> Additionally hooks add more effort for newcomers to the testsuite to grasp it easily.

Typically newcomers will care more about the body of the test rather than setup or teardown.

> and rather keep the scenarios' as simple as possible
> 
> > Yes I agree - my proposal makes them simpler than they currently are, because it avoids polluting scenarios with extra steps for teardown.  For example look at http://benmabey.com/2008/05/19/imperative-vs-declarative-scenarios-in-user-stories.html - the declarative style is much more readable mainly because the scenario is much smaller.
> 
As I mentioned above, if we cover such cases in the `Given` steps (like `Given the image "jeos1" does not exist`), we keep the scenarios clean.

Yes, I like that approach a lot.  As I already said, you may also need to check that quotas are not close to being exceeded, and it would make sense to consider [backgrounds](https://github.com/cucumber/cucumber/wiki/Background) for achieving this.

Thanks for the really good discussion!